File: Easy/733 - FloodFill/Solution.cs
/// Question 733. Flood Fill (https://leetcode-cn.com/problems/flood-fill/submissions/)

public class Solution
{
    /// 执行用时: 324 ms, 在所有 C# 提交中击败了 16.67% 的用户
    /// 内存消耗: 33 MB, 在所有 C# 提交中击败了 100.00% 的用户

    private readonly List<(int, int)> DIRECTIONS = new List<(int, int)>()
    {
        (-1, 0), (1, 0), (0, 1), (0, -1)
    };

    public int[][] FloodFill(int[][] image, int sr, int sc, int newColor)
    {
        int imageRowCount = image.Count();
        int imageColumnCount = image[0].Count();
        if (imageRowCount == 0) { return image; }

        int[][] floodedImage = new int[imageRowCount][];
        for (int i = 0; i < imageRowCount; i++)
        {
            int[] floodedRow = new int[imageColumnCount];
            for (int j = 0; j < imageColumnCount; j++)
            {
                floodedRow[j] = image[i][j];
            }
            floodedImage[i] = floodedRow;
        }

        int startingPixelValue = image[sr][sc];
        floodedImage[sr][sc] = newColor;

        Queue<(int, int)> floodedPixels = new Queue<(int, int)>();
        floodedPixels.Enqueue((sr, sc));

        while (floodedPixels.Any())
        {
            (int currentRow, int currentColumn) = floodedPixels.Dequeue();

            foreach ((int rowDirection, int columnDirection) in this.DIRECTIONS)
            {
                int newRow = currentRow + rowDirection;
                int newColumn = currentColumn + columnDirection;

                if (newRow < imageRowCount && newRow >= 0 &&
                    newColumn < imageColumnCount && newColumn >= 0 &&
                    image[newRow][newColumn] == startingPixelValue &&
                    floodedImage[newRow][newColumn] != newColor)
                {
                    floodedImage[newRow][newColumn] = newColor;
                    floodedPixels.Enqueue((newRow, newColumn));
                }

            }
        }

        return floodedImage;
    }
}

File: Easy/733 - FloodFill/solution.ts
const DIRECTIONS: number[][] = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1],
];

/**
 * 执行用时：104 ms, 在所有 typescript 提交中击败了 36.84% 的用户
 * 内存消耗：39.8 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @returns {number[][]}
 */
function floodFill(
  image: number[][],
  sr: number,
  sc: number,
  newColor: number
): number[][] {
  const imageRow: number = image.length;
  if (imageRow === 0) {
    return image;
  }

  const imageColumn: number = image[0].length;
  const sourcePixel: number = image[sr][sc];
  const floodedImage: number[][] = [];

  for (let i = 0; i < imageRow; i++) {
    let floodedColumn: number[] = [];
    for (let j = 0; j < imageColumn; j++) {
      floodedColumn[j] = image[i][j];
    }
    floodedImage[i] = floodedColumn;
  }

  const floodedPixels: number[][] = [[sr, sc]];
  while (floodedPixels.length) {
    const [pixelRow, pixelColumn] = floodedPixels.shift() as number[];

    for (const [rowDirection, columnDirection] of DIRECTIONS) {
      const newRow: number = pixelRow + rowDirection;
      const newColumn: number = pixelColumn + columnDirection;

      if (
        newRow >= 0 &&
        newRow < imageRow &&
        newColumn >= 0 &&
        newColumn < imageColumn &&
        image[newRow][newColumn] === sourcePixel &&
        floodedImage[newRow][newColumn] !== newColor
      ) {
        floodedPixels.push([newRow, newColumn]);
      }
    }

    floodedImage[pixelRow][pixelColumn] = newColor;
  }

  return floodedImage;
}
